File: Easy/0617-Merge Two Binary Trees/Solution-Iterate.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

typedef struct StackNode{
    struct TreeNode* t1;
    struct TreeNode* t2;
    struct StackNode* next;
}StackNode;

struct TreeNode* mergeTrees(struct TreeNode* t1, struct TreeNode* t2){
    if(t1 == NULL)
        return t2;
    if(t2 == NULL)
        return t1;
    StackNode* top, * push, * pop;
    top = (StackNode*)malloc(sizeof(StackNode));
    top -> next = NULL;
    push = (StackNode*)malloc(sizeof(StackNode));
    push -> t1 = t1;
    push -> t2 = t2;
    push -> next = top;
    top = push;
    while(top -> next != NULL){
        pop = top;
        top = top -> next;
        pop -> t1 -> val += pop -> t2 -> val;
        if(pop -> t1 -> left != NULL && pop -> t2 -> left != NULL){
            push = (StackNode*)malloc(sizeof(StackNode));
            push -> t1 = pop -> t1 -> left;
            push -> t2 = pop -> t2 -> left;
            push -> next = top;
            top = push;
        }else if(pop -> t1 -> left == NULL && pop -> t2 -> left != NULL)
            pop -> t1 -> left = pop -> t2 -> left;
        if(pop -> t1 -> right != NULL && pop -> t2 -> right != NULL){
            push = (StackNode*)malloc(sizeof(StackNode));
            push -> t1 = pop -> t1 -> right;
            push -> t2 = pop -> t2 -> right;
            push -> next = top;
            top = push;
        }else if(pop -> t1 -> right == NULL && pop -> t2 -> right != NULL)
            pop -> t1 -> right = pop -> t2 -> right;
        free(pop);
    }
    free(top);
    return t1;
}

File: Easy/0617-Merge Two Binary Trees/Solution-Iterate.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def mergeTrees(self, t1: TreeNode, t2: TreeNode) -> TreeNode:
        if not t1:
            return t2
        if not t2:
            return t1
        stack = [(t1, t2)]
        while stack:
            n1, n2 = stack.pop()
            n1.val += n2.val
            if n1.left and n2.left:
                stack.append((n1.left, n2.left))
            elif not n1.left and n2.left:
                n1.left = n2.left
            if n1.right and n2.right:
                stack.append((n1.right, n2.right))
            elif not n1.right and n2.right:
                n1.right = n2.right
        return t1
